crates/compiler/constrain/src/expr.rs

```diff
@@ -375,6 +375,38 @@ pub fn constrain_expr(
             let expected_index = expected;
             constraints.equal_types(str_index, expected_index, Category::Str, region)
         }
+        IngestedFile(bytes, anno) => match &anno.typ {
+            Type::Apply(Symbol::STR_STR, _, _) => {
+                if std::str::from_utf8(bytes).is_err() {
+                    todo!("cause an error for the type being wrong due to not being a utf8 string");
+                }
+
+                let str_index = constraints.push_type(types, Types::STR);
+                let expected_index = expected;
+                constraints.equal_types(str_index, expected_index, Category::Str, region)
+            }
+            Type::Apply(Symbol::LIST_LIST, elem_type, _)
+                if matches!(
+                    elem_type[0].value,
+                    Type::DelayedAlias(AliasCommon {
+                        symbol: Symbol::NUM_U8,
+                        ..
+                    })
+                ) =>
+            {
+                let elem_var = Variable::U8;
+                let list_elem_type = Type::Variable(elem_var);
+                let elem_type_index = {
+                    let typ = types.from_old_type(&list_type(list_elem_type));
+                    constraints.push_type(types, typ)
+                };
+                constraints.equal_types(elem_type_index, expected, Category::List, region)
+            }
+            x => todo!(
+                "Unsupported requested type for ingested file, give proper error: {:?}",
```

Comment:
Also, I assume here is also were we somehow need to generate type errors when we request and unsupported type, but not really sure who to do that. Any ideas?